src/mbgl/style/expression/dsl.cpp

```diff
@@ -0,0 +1,147 @@
+#include <mbgl/style/expression/dsl.hpp>
+#include <mbgl/style/expression/literal.hpp>
+#include <mbgl/style/expression/assertion.hpp>
+#include <mbgl/style/expression/coercion.hpp>
+#include <mbgl/style/expression/equals.hpp>
+#include <mbgl/style/expression/interpolate.hpp>
+#include <mbgl/style/expression/compound_expression.hpp>
+#include <mbgl/util/ignore.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+namespace dsl {
+
+template <class... Args>
+static std::vector<std::unique_ptr<Expression>> vec(Args... args) {
+    std::vector<std::unique_ptr<Expression>> result;
+    util::ignore({ (result.push_back(std::move(args)), 0)... });
+    return result;
+}
+
+template <class... Args>
+static std::unique_ptr<Expression> compound(const char* op, Args... args) {
+    ParsingContext ctx;
+    ParseResult result =  createCompoundExpression(op, vec(std::move(args)...), ctx);
+    assert(result);
+    return std::move(*result);
+}
+
+std::unique_ptr<Expression> literal(const char* value) {
+    return literal(std::string(value));
+}
+
+std::unique_ptr<Expression> literal(Value value) {
+    return std::make_unique<Literal>(value);
+}
+
+std::unique_ptr<Expression> number(std::unique_ptr<Expression> value) {
+    return std::make_unique<Assertion>(type::Number, vec(std::move(value)));
+}
+
+std::unique_ptr<Expression> string(std::unique_ptr<Expression> value) {
+    return std::make_unique<Assertion>(type::String, vec(std::move(value)));
+}
+
+std::unique_ptr<Expression> toColor(const char* value) {
+    return toColor(literal(value));
+}
+
+std::unique_ptr<Expression> toColor(std::unique_ptr<Expression> value) {
+    return std::make_unique<Coercion>(type::Color, vec(std::move(value)));
+}
+
+std::unique_ptr<Expression> get(const char* value) {
+    return get(literal(value));
+}
+
+std::unique_ptr<Expression> get(std::unique_ptr<Expression> property) {
+    return compound("get", std::move(property));
+}
+
+std::unique_ptr<Expression> id() {
+    return compound("id");
+}
+
+std::unique_ptr<Expression> zoom() {
+    return compound("zoom");
+}
+
+std::unique_ptr<Expression> eq(std::unique_ptr<Expression> lhs,
+                               std::unique_ptr<Expression> rhs) {
+    return std::make_unique<Equals>(std::move(lhs), std::move(rhs), false);
```

Comment:
This and the other non-compound-expression-based DSL functions bypass type validation.  Since the DSL now lowers the barrier to creating expressions directly, without parsing/validation, I think it's worthwhile to include some `assert`s on the input expressions' types in the various expression constructors.